File: backend/routers/benchmarks.py
from fastapi import APIRouter, HTTPException, Depends
from typing import Optional
from database import crud
from dependencies import require_admin

router = APIRouter(prefix="/api/benchmarks", tags=["benchmarks"])


@router.get("/runs")
async def list_benchmark_runs(
    limit: int = 50,
    dataset: Optional[str] = None,
    provider: Optional[str] = None,
    current_user: dict = Depends(require_admin),
):
    """List benchmark runs with optional filters."""
    _ = current_user
    limit = max(1, min(limit, 200))
    return await crud.list_benchmark_runs(
        limit=limit,
        dataset=dataset,
        provider=provider,
    )


@router.get("/runs/{run_id}")
async def get_benchmark_run(
    run_id: int,
    current_user: dict = Depends(require_admin),
):
    """Get a benchmark run by ID."""
    _ = current_user
    run = await crud.get_benchmark_run(run_id)
    if not run:
        raise HTTPException(status_code=404, detail="Benchmark run not found")
    return run


@router.get("/runs/{run_id}/results")
async def get_benchmark_results(
    run_id: int,
    current_user: dict = Depends(require_admin),
):
    """Get all results for a benchmark run."""
    _ = current_user
    run = await crud.get_benchmark_run(run_id)
    if not run:
        raise HTTPException(status_code=404, detail="Benchmark run not found")
    results = await crud.get_benchmark_results(run_id)
    return results


@router.get("/compare")
async def compare_models(
    dataset: str = "cord",
    limit: int = 20,
    current_user: dict = Depends(require_admin),
):
    """Get comparison of all models on a dataset."""
    _ = current_user
    limit = max(1, min(limit, 200))
    comparison = await crud.get_model_comparison(dataset=dataset, limit=limit)

    if not comparison:
        return {"message": "No benchmark data available", "runs": []}

    runs = []
    for row in comparison:
        runs.append(
            {
                "provider": row["provider"],
                "model": row["model"],
                "processing_method": "vision",
                "sample_count": None,
                "overall_accuracy": row["avg_accuracy"],
                "avg_latency": None,
                "total_cost": None,
                "total_prompt_tokens": None,
                "total_completion_tokens": None,
                "success_rate": None,
                "started_at": None,
            }
        )

    return {"runs": runs[:limit]}


@router.get("/models")
async def get_benchmarked_models(
    current_user: dict = Depends(require_admin),
):
    """Return all provider/model combos that have benchmark data."""
    _ = current_user
    return await crud.get_benchmarked_models_summary()

File: frontend/src/lib/api/benchmarks.ts
import { API_BASE, getAuthHeaders, parseApiError } from './client';
import type { BenchmarkRun, BenchmarkResult, ModelComparison, BenchmarkedModel, UsageAnalytics } from './types';

export async function listBenchmarkRuns(limit = 50, dataset?: string, provider?: string): Promise<BenchmarkRun[]> {
  const params = new URLSearchParams();
  params.append('limit', limit.toString());
  if (dataset) params.append('dataset', dataset);
  if (provider) params.append('provider', provider);

  const response = await fetch(`${API_BASE}/benchmarks/runs?${params}`, {
    headers: getAuthHeaders()
  });
  if (!response.ok) {
    throw await parseApiError(response, 'Failed to load benchmark runs');
  }
  return response.json();
}

export async function getBenchmarkRun(runId: number): Promise<BenchmarkRun> {
  const response = await fetch(`${API_BASE}/benchmarks/runs/${runId}`, {
    headers: getAuthHeaders()
  });
  if (!response.ok) {
    throw await parseApiError(response, 'Benchmark run not found');
  }
  return response.json();
}

export async function getBenchmarkResults(runId: number): Promise<BenchmarkResult[]> {
  const response = await fetch(`${API_BASE}/benchmarks/runs/${runId}/results`, {
    headers: getAuthHeaders()
  });
  if (!response.ok) {
    throw await parseApiError(response, 'Failed to load benchmark results');
  }
  return response.json();
}

export async function compareModels(dataset = 'cord', limit = 20): Promise<{ runs: ModelComparison[] }> {
  const params = new URLSearchParams();
  params.append('dataset', dataset);
  params.append('limit', limit.toString());

  const response = await fetch(`${API_BASE}/benchmarks/compare?${params}`, {
    headers: getAuthHeaders()
  });
  if (!response.ok) {
    throw await parseApiError(response, 'Failed to load benchmark comparison');
  }
  return response.json();
}

export async function getBenchmarkedModels(): Promise<BenchmarkedModel[]> {
  const response = await fetch(`${API_BASE}/benchmarks/models`, {
    headers: getAuthHeaders(),
  });
  if (!response.ok) {
    return [];
  }
  return response.json();
}

export async function getUsageAnalytics(filters?: {
  date_from?: string;
  date_to?: string;
  provider?: string;
  model?: string;
  schema_name?: string;
  processing_method?: string;
  document_type?: string;
}): Promise<UsageAnalytics> {
  const params = new URLSearchParams();
  Object.entries(filters || {}).forEach(([key, value]) => {
    if (value) params.append(key, value);
  });

  const response = await fetch(`${API_BASE}/analytics/usage?${params.toString()}`, {
    headers: getAuthHeaders()
  });
  if (!response.ok) {
    throw await parseApiError(response, 'Failed to load usage analytics');
  }
  return response.json();
}
